web/src/chat/MessageItem.js

```diff
@@ -33,24 +34,83 @@ const MessageItem = ({
   onRegenerate,
   onLike,
   onToggleRead,
+  onEditMessage,
   disableInput,
   isReading,
   readingMessage,
   sendMessage,
 }) => {
   const [avatarSrc, setAvatarSrc] = useState(null);
+  const [isEditing, setIsEditing] = useState(false);
+  const [editedText, setEditedText] = useState("");
+  const [isHovering, setIsHovering] = useState(false);
 
   useEffect(() => {
     // Set the initial avatar source
     setAvatarSrc(message.author === "AI" ? avatar : Setting.getUserAvatar(message, account));
   }, [message.author, avatar, account, message]);
 
+  const handleEditActions = {
+    start: () => {
+      setIsEditing(true);
+      setEditedText(message.text);
+    },
+    save: () => {
+      onEditMessage({...message, text: editedText, updatedTime: new Date().toISOString()});
+      setIsEditing(false);
+    },
+    cancel: () => {
+      setIsEditing(false);
+      setEditedText("");
+    },
+    keyDown: (e) => {
+      if (e.key === "Enter" && !e.shiftKey) {
+        e.preventDefault();
+        handleEditActions.save();
+      }
+    },
+  };
+
   const handleAvatarError = () => {
     // Set fallback URL when avatar fails to load
     setAvatarSrc(AvatarErrorUrl);
   };
 
+  const renderEditForm = () => (
```

Comment:
Put this UI widget into a new JS file as a new React component:

![image](https://github.com/user-attachments/assets/be9dfc8b-591a-4203-896a-5223f906604c)